docs/discovery-naming-conventions.md

```diff
@@ -0,0 +1,137 @@
+   onto the next _Component_ and shorten using the _short hash_. Continue until
+   the resulting _Discovered Name_ is 63 characters or less.
+
+5. In the case that the _Discovered Name_ is longer than 63 characters after the
+   shortening process, the SHA256 hash produced in step 2 will be used as the
+   _Discovered Name_, prepended with `g-` to ensure the name starts with a
+   letter. This scenario will only happen if *all* _Components_ of the initial
+   _Discovered Name_ are longer than 21 characters, which is a rare case.
+
+#### Example
+
+- `${discoverer-prefix}`: `k8s`
+- `${service-name}`: `the-really-long-kube-service-name-that-is-exactly-63-characters`
+- `${backend-name}`: `us-east-cluster`
+
+1. The _Discovered Name_ has a total of three components. Thus, allocate 21 (61/3) characters to each.
+
+2. Take the SHA256 hash of `k8s-the-really-long-kube-service-name-that-is-exactly-63-characters-us-east-cluster`.
+
+    ```
+    SHA256 hash = cdbeb21c3261255a91ded1844754791ef1ec06ecbe5b2421ff2abe221950fc72
+    Short hash = cdbeb2
+    ```
+
+3. Check if the last _Component_ of the _Discovered Name_ goes over 21 characters:
+
+    ```
+    "us-east-cluster" has 15 characters. Move onto next component.
+    ```
+
+4. Check if the next _Component_ goes over 21 characters:
+
+    ```
+    "the-really-long-kube-service-name-that-is-exactly-63-characters" has 63 characters.
+    Shorten using the short hash by truncating and appending the short hash.
+    Result: "the-really-lon-cdbeb2".
+    ```
+
+5. Check if the resulting _Discovered Name_ is shorter than 63 characters:
+
+    ```
+    "k8s-the-really-lon-cdbeb2-us-east-cluster" has 41 characters.
+    Thus, we have arrived at our shortened Discovered Name.
+    ```
+
+Discovered name after shortening: `k8s-the-really-lon-cdbeb2-us-east-cluster`
+
+## Discoverer Specifics
+
+The specifics of each discoverer are documented below.
+
+### Kubernetes discoverer
+
+- `${discoverer-prefix}`: `k8s`
+- `${service-name}`: The name of the backend service, verbatim.
+- `${backend-name}`: The value of the `--cluster-name` flag provided to the
+  discoverer.
+
+Service port names are copied verbatim from the backend service.
+
+### OpenStack discoverer
+
+- `${discoverer-prefix}`: `os`
+- `${service-name}`: `${name}-${id}` of the LBaaS Load Balancer. Both are
+  lowercased during the discovery process.
+- `${backend-name}`: The value of the `--cluster-name` flag provided to the
+  discoverer.
+
+Service port names are set to `${discoverer-prefix}-${port-number}`.
+
+#### Why is the `${service-name}` a composite name?
+
+The `${service-name}` produced by the OpenStack discoverer is composed of the
+name and ID of the LBaaS Load Balancer. This is required because names are not
+guaranteed to be unique in an OpenStack project.
+
+#### What happens to Load Balancers that do not have a name?
+
+Names in OpenStack are optional. In this scenario, the `${service-name}` will
```

Comment:
Should this be error and thrown out in the same way invalid characters make it error out?

Reply:
We could certainly do this, I am just not sure about how common this is in openstack world.

Reply:
I don't think it is, but just helps to make the ui cleaner, so a `kubectl get svc` would have all services grouped by type. Just thinking out loud.

Reply:
I don't see why this needs to be an error and thrown out.  From a k8s perspective, there's nothing wrong with the $service-name being just the uuid?  [I guess it would be `-$uuid`].

Reply:
This is probably a more general issue then -- in Kubernetes, object names are immutable.  I'm assuming that's not the case in OpenStack (and probably not other backends).  If I rename an OpenStack LBaaS, does that break any routes I've defined using the discovered service?

Reply:
@rosskukulinski I think it would indeed break any routes you have defined with the discovered service. The discoverer would delete the existing service, and then recreate it with the new name.